src/core/lib.lua

```diff
@@ -686,6 +686,38 @@ function getenv (name)
    else return nil end
 end
 
+function fstruct(def)
+   local struct = {}
+   local offset = 0
+   for ct, fld in def:gmatch('([%a_][%w_]*)%s+([%a_][%w_]*);') do
+      ct = ffi.typeof(ct)
+      struct[fld] = {
+         fieldname = fld,
+         ct = ct,
+         size = ffi.sizeof(ct),
+         offset = offset,
+      }
+      offset = offset + struct[fld].size
+   end
+   return struct, offset
+end
+
+function fieldrd(field, fd)
+   local buf = ffi.typeof('$ [1]', field.ct)()
+   local r, err = fd:pread(buf, field.size, field.offset)
+   if not r then error(err) end
+   return buf[0]
+end
+
+function fieldwr(field, fd, val)
+   local buf = ffi.typeof('$ [1]', field.ct)()
```

Comment:
This function is a source of a trace abort.

e.g. in a -jv output there are these lines

```
[TRACE --- (20/0) zone.lua:38 -- NYI: unsupported variant of FastFunc ffi.typeof at lib.lua:713]
[TRACE --- (20/0) zone.lua:38 -- NYI: unsupported variant of FastFunc ffi.typeof at lib.lua:713]
[TRACE --- (20/0) zone.lua:38 -- NYI: unsupported variant of FastFunc ffi.typeof at lib.lua:713]
[TRACE --- (20/0) zone.lua:38 -- NYI: unsupported variant of FastFunc ffi.typeof at lib.lua:713]
[TRACE  22 (20/0) zone.lua:38 -- fallback to interpreter]
```

AFAIU This means that a trace 22, coming from trace 20, met something it couldn't compile, so it is falling back to the interpreter.  I don't know why it printed it like 4 times, did it try 4 times?  I don't know.  Probably so, thinking maybe if it tried again the trace would go somewhere else.  It means that this allocation won't be sunk, the whole thing will go on the interpreted path, etc.  It's only used in one place; let's just inline it into its caller.  Same with fstruct and fieldrd.